### DISCUSSION TITLE
Inverter Power Register X3 MIC - smart meter emulator

### DISCUSSION THREAD

**Discussion by raenji-sk** [wills106/homeassistant-solax-modbus#923] [General]
@wills106 Do you know what the registers 0x435 and 0x43B represent? In the source code for Solax they are named inverter power.
I am trying to emulate a smart meter and after the inverter asks for meter type and I send a correct response it asks for other registers.
Namely 14, 40, 62 and 92 from the meter. 62 and 92 are not required according to several sources, 14 (1 register, 2 bytes int16) should be grid import or export according to https://github.com/syssi/esphome-solax-x1-mini/issues/20.
But what is register 40? It asks for two registers, 4 bytes, probably UInt32...
When I send reply to it the value of the sensor in your integration named Inverter Power starts to output some values but they do not seem to be correct. For example I have PV generating 1500W but inverter power states 4000-5000W. 

So the question is what does this sensor represent and do you have any idea how it is calculated? 

Or does perhaps anyone have a list of register used by the Solax smart meters with their firmware installed?

**Comment by Kars-de-Jong** [comment]
Hi @raenji-sk,

There are some issues with the X3 MIC Gen1 handling in this integration at the moment.

0x435 is not inverter power, it is Pgrid ("Measured Power"). I am going to create a PR to fix this.
Basically everything that is now called "Measured Power" should be "Inverter Power". That's also the reason why the "Grid Import" and "Grid Export" sensors have the wrong values.

I am emulating a different meter (a single phase Solax SDM230, based on https://github.com/syssi/esphome-solax-x1-mini/issues/20#issuecomment-1118029392), which helped me to find out what was wrong.

On the SolaX side:
0x435 = 2 holding registers with measured power (signed int32) in units of W, negative means importing, positive means exporting
0x437 = 2 holding registers with total grid export, in units of 0.01 kwh (unsigned int32)
0x439 = 2 holding registers with total grid import, in units of 0.01 kwh (unsigned int32)
0x43b = 1 holding register with today's grid export, in units of 0.01 kwh (unsigned int16)
0x43d = 1 holding register with today's grid import, in units of 0.01 kwh (unsigned int16)

On the meter side:
0xb = 1 holding register with meter type, value 168 (0xa8)
0xb = 2 input registers with measured power (float) in units of W, positive means importing, negative means exporting
0x48 = 2 input registers with total grid import (float) in units of kWh
0x4a = 2 input registers with total grid export (float) in units of kWh
